### PR TITLE
feature/typed strategies and encryption_key_arn support on create_memory

### PR DESCRIPTION
**Pull request by jona62**
## Description

This PR introduces **typed memory strategies** to the Bedrock AgentCore Memory Manager, providing a strongly-typed, Pydantic-based interface for creating and managing memory strategies. The implementation adds support for encryption key ARNs and maintains full backward compatibility with existing dictionary-based strategy configurations.

### Key Features Added:

1. **Typed Strategy Classes**: New Pydantic-based strategy classes (`SemanticStrategy`, `SummaryStrategy`, `UserPreferenceStrategy`, `CustomSemanticStrategy`)
2. **Configuration Models**: `ExtractionConfig` and `ConsolidationConfig` for customizable memory operations
3. **Encryption Support**: Added `encryption_key_arn` parameter to memory creation methods
4. **Backward Compatibility**: Seamless integration with existing dictionary-based strategies
5. **Type Safety**: Full type hints and validation using Pydantic models

#### Creating Typed Strategies
```python
from bedrock_agentcore_starter_toolkit.operations.memory.models import (
    SemanticStrategy, CustomSemanticStrategy, ExtractionConfig, ConsolidationConfig
)

# Simple semantic strategy
semantic = SemanticStrategy(
    name="ConversationSemantics",
    description="Extract semantic information from conversations",
    namespaces=["semantics/{actorId}/{sessionId}"]
)

# Custom strategy with specific model configurations
custom = CustomSemanticStrategy(
    name="CustomExtraction",
    extraction_config=ExtractionConfig(
        append_to_prompt="Extract key business insights",
        model_id="anthropic.claude-3-sonnet-20240229-v1:0"
    ),
    consolidation_config=ConsolidationConfig(
        append_to_prompt="Consolidate business insights",
        model_id="anthropic.claude-3-haiku-20240307-v1:0"
    )
)
```

#### Using Typed Strategies with Memory Manager
```python
# Create memory with typed strategies
memory = manager.create_memory_and_wait(
    name="TypedMemory",
    strategies=[semantic, custom],
    encryption_key_arn="arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012"
)

# Mixed usage (typed + legacy dict strategies)
mixed_strategies = [
    SemanticStrategy(name="TypedStrategy"),
    {"summaryMemoryStrategy": {"name": "LegacyStrategy"}}  # Still supported
]
```

#### Strategy Conversion System
```python
from bedrock_agentcore_starter_toolkit.operations.memory.models import convert_strategies_to_dicts

# Automatic conversion handles both typed and dict strategies
strategies = [semantic_strategy, legacy_dict_strategy]
api_ready_strategies = convert_strategies_to_dicts(strategies)
```

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

## Testing

- [x] Unit tests pass locally
- [x] Integration tests pass (if applicable)
- [x] Test coverage remains above 80%
- [x] Manual testing completed

### Test Coverage:
- **545 new test cases** added in `test_strategy_types.py`
- Comprehensive validation testing for all strategy types
- Pydantic model serialization and validation tests
- Backward compatibility testing with mixed strategy types
- Error handling and edge case validation

## Checklist

- [x] My code follows the project's style guidelines (ruff/pre-commit)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published

## Security Checklist

- [x] No hardcoded secrets or credentials
- [x] No new security warnings from bandit
- [x] Dependencies are from trusted sources
- [x] No sensitive data logged

## Breaking Changes

**None** - This implementation maintains full backward compatibility:

- Existing dictionary-based strategy configurations continue to work unchanged
- All existing method signatures remain the same (new parameters are optional)
- The `convert_strategies_to_dicts()` function handles mixed strategy types seamlessly
- Legacy code requires no modifications

### Architecture Decisions:

1. **Pydantic Integration**: Chose Pydantic for robust validation, serialization, and IDE support
2. **Abstract Base Class**: `BaseStrategy` ensures consistent interface across all strategy types
3. **Conversion Layer**: `convert_strategies_to_dicts()` provides seamless backward compatibility
4. **Type Union**: `StrategyType` union allows mixed usage of typed and dictionary strategies

### Files Modified:
- `manager.py`: Enhanced with typed strategy support and encryption key parameter
- `models/__init__.py`: New module with strategy classes and conversion utilities
- `models/strategies/`: New strategy implementation modules
- `test_strategy_types.py`: Comprehensive test suite (545 test cases)
- `memory-manager.ipynb`: Updated integration examples

### Performance Impact:
- Minimal overhead from Pydantic validation
- No impact on existing dictionary-based workflows
- Type checking provides better development experience and fewer runtime errors

This enhancement significantly improves the developer experience while maintaining full compatibility with existing implementations.

### CODE REVIEW

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+from pydantic import Field
+
+from .base import BaseStrategy, ConsolidationConfig, ExtractionConfig
+
+
+class CustomSemanticStrategy(BaseStrategy):
+    """Custom semantic strategy with configurable extraction and consolidation.
+
+    This strategy allows full customization of both extraction and consolidation
```
Comment (bergjaak) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py]: We should just say: Allows tuning extraction parameters and consolidation parameters and providing your own bedrock LLM and prompt suffix

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/semantic.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        config = {
+            "name": self.name,
+        }
+
+        if self.description is not None:
+            config["description"] = self.description
+
+        if self.namespaces is not None:
+            config["namespaces"] = self.namespaces
+        else:
```
Comment (bergjaak) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/semantic.py]: don't we also default on the server? Should we not have this `else` because of that?

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/summary.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        config = {
+            "name": self.name,
+        }
+
+        if self.description is not None:
+            config["description"] = self.description
+
+        if self.namespaces is not None:
+            config["namespaces"] = self.namespaces
+        else:
```
Comment (bergjaak) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/summary.py]: we default on server, right? On CreateMemory

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/user_preference.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        config = {
+            "name": self.name,
+        }
+
+        if self.description is not None:
+            config["description"] = self.description
+
+        if self.namespaces is not None:
+            config["namespaces"] = self.namespaces
+        else:
```
Comment (bergjaak) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/user_preference.py]: we default on server, right? On CreateMemory

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                }
+            },
+        }
+
+        if self.description is not None:
+            config["description"] = self.description
+
+        if self.namespaces is not None:
+            config["namespaces"] = self.namespaces
+        else:
```
Comment (bergjaak) [inline src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py]: we default on server, right? On CreateMemory

Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py]: Removed

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/semantic.py (hunk 2 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/semantic.py]: Removed

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/summary.py (hunk 3 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/summary.py]: Removed

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/user_preference.py (hunk 4 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/user_preference.py]: Removed

File: src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py (hunk 1 above)
Comment (jona62) [reply src/bedrock_agentcore_starter_toolkit/operations/memory/models/strategies/custom.py]: Removed "full"

**Review by bergjaak — APPROVED** [review]
